codegen/value.go

```diff
@@ -208,6 +217,55 @@ func (v *errorValue) Reflect(t reflect.Type) (reflect.Value, error) {
 	return reflect.Value{}, v.err
 }
 
+type lazyValue struct {
+	valCh <-chan Value
+	once  *sync.Once
+	done  chan struct{}
+	val   Value
+}
+
+func (v *lazyValue) wait() {
+	v.once.Do(func() {
+		v.val = <-v.valCh
+		close(v.done)
+	})
```

Comment:
Another way to do this:

```suggestion
	val, ok := <-v.valCh:
	if ok {
		v.val = val
		close(v.valCh)
	}
```

Then you don't need `once` or `done`.

Reply:
If there are multiple goroutines waiting on `<-v.valCh` is it possible a context switch happens after line 228, and calls `(*lazyValue).Filesystem` for example, causing a nil pointer dereference before line 230 `v.val = val` is assigned?

Reply:
That is not possible. Only one goroutine will receive a value from the channel (`ok == true`). The rest will block until the `close` on line 231, and receive `ok == false`.

Reply:
Ah makes sense.